feat: Add properties to SSI Phrase Builder analytics events (M2-8295)

farmerpaul opened the pull request:
- [x] Tests for the changes have been added

### 📝 Description

<!-- Contributions are welcome! If there is a corresponding      -->
<!-- JIRA ticket, link to it by replacing `#` with ticket number -->

🔗 [Jira Ticket M2-8295](https://mindlogger.atlassian.net/browse/M2-8295)

This PR adds new SSI-related properties to these existing Mixpanel analytics events:
- `Applet Save click`
- `Applet edit successful`
- `Applet Created Successfully`

The new properties are:

> [!NOTE]
> For all item counts below, only items that are **not hidden** should be captured in the counts. Hidden items are not of interest to the data team and should be ignored.

- `Item Count`: Total count of all items in the applet
- `Phrase Builder Item Count`: Total count of Phrase Builder items in the applet (𝑥)
- `Count of Unique Items in Phrase Builders`: Sum of the counts of the unique items referenced within each Phrase
Builder item in the applet (𝑦)
- `Average Unique Items Per Phrase Builder`: The result of property (𝑦) divided by property (𝑥). *
- `Average Items Per Phrase Builder`: The number of "Item (Response)" fields contained in all Phrase Builders divided by property (𝑥). *
- `Average Text Boxes per Phrase Builder`: The number of "Text" fields contained in all Phrase Builders divided by property (𝑥). *
- `Average Line Breaks per Phrase Builder`: The number of "Line Break" fields contained in all Phrase Builders divided by property (𝑥). *

> [!NOTE]
> \* For any of these properties, where the result is a division by the # of Phrase Builders (𝑥): if (𝑥) is zero, then the value of the property should be null (not set).

### 🪤 Peer Testing

Refer to the [ticket description](https://mindlogger.atlassian.net/browse/M2-8295) for a detailed listing of how to trigger the relevant events and what properties should be attached to them.

**For each event:**

1. Follow the instructions under the **Trigger** column to trigger the event.
2. Open Mixpanel, and navigate to the Events tab.
3. Refresh if needed.
    **Expected outcome:** The event matching the **Event Name** column should have been added to the top of the list.
4. Expand the event and open the **Your Properties** tab:
    **Expected outcome:** The custom properties for that event mentioned in the **Details** column should be listed.

farmerpaul committed:
feat: add additional SSI properties to analytics

farmerpaul committed:
test: update/add unit tests

farmerpaul commented:
@sultanofcardio of the PRs I've requested a review from you for, this one should be the priority. Thx! 🙏🙏

sultanofcardio commented on src/shared/utils/mixpanel/mixpanel.utils.ts:
```diff
@@ -26,19 +26,49 @@ export const trackAppletSave = ({
+    [MixpanelProps.PhraseBuilderItemCount]: phraseBuilderItemCount,
+    [MixpanelProps.ItemsIncludedInPhraseBuilders]: itemsIncludedInPhraseBuilders,
+    [MixpanelProps.AverageItemsPerPhraseBuilder]: phraseBuilderItemCount
+      ? itemsIncludedInPhraseBuilders / phraseBuilderItemCount
```
This calculation can result in long decimals, which may not be desirable. We could do something like this instead

```suggestion
      ? parseFloat((itemsIncludedInPhraseBuilders / phraseBuilderItemCount).toFixed(2))
```

farmerpaul committed:
fix: correction to item count math

The `Items Included in Phrase Builders` property should actually be a
sum of the counts of the unique items referenced within each Phrase
Builder, rather than a count of the unique items referenced across all
Phrase Builders.

farmerpaul committed:
fix: round average down to 2 decimal places

farmerpaul committed:
test: fix tests

farmerpaul committed:
feat: support new SSI properties

The data team has requested tracking new SSI properties in Mixpanel,
as well as requested that hidden items be ignored from all counts.

farmerpaul committed:
test: update unit tests

farmerpaul commented:
@sultanofcardio could I ask you to review this once again? The data team has requested adding some new properties, and to skip hidden items from item counts – please see description above. Thank you! 🙏🏻

sultanofcardio commented on src/shared/utils/mixpanel/mixpanel.utils.ts:
```diff
@@ -27,40 +27,98 @@ export const trackAppletSave = ({
     }
+
+    const visibleItemNames = new Set<string>();
+    const referencedItemNames = new Set<string>();
```
I think this `referencedItemNames` is a remnant of the previous implementation. It's shadowed in the inner loop and doesn't end up doing anything, so we can remove it

farmerpaul replied:
Ah yes, the result of my not doing a careful enough rebase!

sultanofcardio commented on src/shared/utils/mixpanel/mixpanel.utils.ts:
```diff
@@ -27,40 +27,98 @@ export const trackAppletSave = ({
+            }
+          }
+        }
+        itemsIncludedInPhraseBuilders += referencedItemNames.size;
```
What's being counted here is the _sum of the counts of the unique items referenced within each Phrase
Builder_, like you mention in f21c30e0. I trust that this is what it should actually be, but it's a little different from what the ticket description says

> Total count of unique items that were included as Item Response fields in Phrase Builder items in the applet

I can see that being misinterpreted by QA as what you had implemented before: _count of the unique items referenced across all Phrase Builders_

So we should probably update the ticket description.

Moreover, I think calling this `Items Included in Phrase Builders` is misleading. Given what it actually is, something like `Phrase Builder Unique Item Response Count` (or something like that) is more appropriate

farmerpaul replied:
I've updated the ticket description to redefine that parameter, and renamed it as suggested!

sultanofcardio commented on src/shared/utils/mixpanel/mixpanel.utils.ts:
```diff
@@ -27,40 +27,98 @@ export const trackAppletSave = ({
   let manualAssignedActivityCount = 0;
-  for (const activity of applet.activities ?? []) {
-    itemTypes.push(...(activity.items ?? []).map((item) => item.responseType));
+  const itemTypes: ItemResponseType[] = [];
```
This array seems to only be used as the argument to a Set constructor. Maybe we could just make this a Set

sultanofcardio reviewed (approved):
A few non blocking comments, but looks good ✅

farmerpaul committed:
fix: remove unused code from bad rebase

farmerpaul committed:
fix: define `itemTypes` to just be a Set

farmerpaul committed:
fix: rename event property
